### PR TITLE
feat[next]: Enable tests for embedded with cupy

### PR DESCRIPTION
**Pull request by havogt**
Introduces mechanism in tests for having different allocators for the same (`None`) backend.

Fixes:
- The resulting buffer for scan is deduced from the buffer type of the arguments, if there are no arguments we fallback to numpy (maybe break). We need to find a mechanism for this corner case. Currently these tests are excluded with `pytest.mark.uses_scan_without_field_args` for cupy embedded execution.

Refactoring:
- make common.field and common.connectivity private
- rename next_tests.exclusion_matrices to definitions

TODOs for later:
- `broadcast` of scalar ignores the broadcast

### CODE REVIEW

File: tests/next_tests/integration_tests/cases.py [hunk 1]
```diff
@@ -482,17 +482,18 @@ def verify_with_default_data(
 @pytest.fixture
 def cartesian_case(fieldview_backend):  # noqa: F811 # fixtures
     yield Case(
-        fieldview_backend,
+        fieldview_backend if isinstance(fieldview_backend, ppi.ProgramExecutor) else None,
```
Comment (havogt) [inline tests/next_tests/integration_tests/cases.py:485]: this is weird and dangerous, let's discuss a better strategy

File: src/gt4py/next/common.py [hunk 2]
```diff
@@ -844,7 +844,7 @@ def is_connectivity_field(
 
 
 @functools.singledispatch
-def field(
+def _field(
```
Comment (havogt) [inline src/gt4py/next/common.py:847]: add comment for what this is used

File: src/gt4py/next/embedded/operators.py [hunk 3]
```diff
@@ -141,10 +150,21 @@ def _intersect_scan_args(
...
 
 
-def _construct_scan_array(domain: common.Domain):
+def _get_array_ns(
+    *args: core_defs.Scalar | common.Field | tuple[core_defs.Scalar | common.Field | tuple, ...]
+) -> ModuleType:
+    for arg in utils.flatten_nested_tuple(args):
+        if hasattr(arg, "array_ns"):
+            return arg.array_ns
+    return np
```
Comment (havogt) [inline src/gt4py/next/embedded/operators.py:159]: instead of returning np, maybe the save thing would be to error for now.

File: docs/development/ADRs/0015-Test_Exclusion_Matrices.md [hunk 4]
```diff
@@ -1,13 +1,13 @@
 ---
-tags: []
+tags: [testing]
 ---
 
 # Test-Exclusion Matrices
 
 - **Status**: valid
 - **Authors**: Edoardo Paone (@edopao), Enrique G. Paredes (@egparedes)
```
Comment (egparedes) [inline docs/development/ADRs/0015-Test_Exclusion_Matrices.md:8]: Should you add your name here somehow?

File: src/gt4py/next/common.py [hunk 5]
```diff
@@ -844,18 +844,20 @@ def is_connectivity_field(
...
     definition: Any,
     /,
     *,
     domain: Optional[DomainLike] = None,
     dtype: Optional[core_defs.DType] = None,
 ) -> Field:
+    # Utility function to construct a `Field` from different buffer representations.
+    # Consider removing this function and using `Field` constructor directly.
```
Comment (egparedes) [inline src/gt4py/next/common.py:855]: Put the comment above the function (and optionally mention also `_connectivity`)

File: tests/next_tests/definitions.py [hunk 6]
```diff
@@ -55,6 +59,32 @@ class ProgramBackendId(_PythonObjectIdMixin, str, enum.Enum):
...
+
+    @property
+    def allocator(self) -> next_allocators.FieldBufferAllocatorProtocol:
+        ...
+
+
+@dataclasses.dataclass(frozen=True)
+class EmbeddedExecutionDescriptor:
+    allocator: next_allocators.FieldBufferAllocatorProtocol
+    executor: NoneType = dataclasses.field(default=None, init=False)
```
Comment (egparedes) [inline tests/next_tests/definitions.py:76]: I think it should be already enough to define the executor at the class level.
```suggestion
    executor: Final = None
```

**Review by egparedes — COMMENTED** [review]
Minor comments and questions. It should be ready to merge in the next iteration.

File: docs/development/ADRs/0015-Test_Exclusion_Matrices.md (hunk 4 above)
Comment (egparedes) [reply docs/development/ADRs/0015-Test_Exclusion_Matrices.md:8]: :+1:
